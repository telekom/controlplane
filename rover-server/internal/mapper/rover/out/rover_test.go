// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package out

import (
	"github.com/gkampitakis/go-snaps/match"
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"github.com/telekom/controlplane/rover-server/internal/api"
)

var _ = Describe("Rover Mapper", func() {
	Context("MapRover", func() {
		It("must map the internal format to the api format", func() {
			input := rover
			output := &api.Rover{}

			err := MapRover(input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must return an error if the input rover is nil", func() {
			output := &api.Rover{}

			err := MapRover(nil, output)

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("input rover is nil"))
		})

	})

	Context("MapExposures", func() {
		It("must map exposures correctly", func() {
			input := rover
			output := &api.Rover{}

			err := mapExposures(input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("MapSubscriptions", func() {
		It("must map subscriptions correctly", func() {
			input := rover
			output := &api.Rover{}

			err := mapSubscriptions(input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("MapRoverResponse", func() {
		It("must map a Rover to a RoverResponse correctly", func() {
			input := GetRoverWithReadyCondition(rover)

			output, err := MapRoverResponse(ctx, input)

			Expect(err).To(BeNil())

			Expect(output).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), output, match.Any("status.time"))
		})

		It("must return an error if the input rover is nil", func() {
			output, err := MapRoverResponse(ctx, nil)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("input rover is nil"))
		})

	})
})
