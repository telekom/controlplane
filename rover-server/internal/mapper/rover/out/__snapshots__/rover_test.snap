
[Rover Mapper MapRover must map the internal format to the api format - 1]
&api.Rover{
    Authentication: api.Authentication{},
    Exposures:      {
        {
            union: {0x7b, 0x22, 0x61, 0x70, 0x70, 0x72, 0x6f, 0x76, 0x61, 0x6c, 0x22, 0x3a, 0x22, 0x53, 0x49, 0x4d, 0x50, 0x4c, 0x45, 0x22, 0x2c, 0x22, 0x62, 0x61, 0x73, 0x65, 0x50, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x65, 0x6e, 0x69, 0x2f, 0x64, 0x69, 0x73, 0x74, 0x72, 0x2f, 0x76, 0x31, 0x22, 0x2c, 0x22, 0x74, 0x79, 0x70, 0x65, 0x22, 0x3a, 0x22, 0x61, 0x70, 0x69, 0x22, 0x2c, 0x22, 0x75, 0x70, 0x73, 0x74, 0x72, 0x65, 0x61, 0x6d, 0x22, 0x3a, 0x22, 0x68, 0x74, 0x74, 0x70, 0x73, 0x3a, 0x2f, 0x2f, 0x68, 0x74, 0x74, 0x70, 0x62, 0x69, 0x6e, 0x2e, 0x6f, 0x72, 0x67, 0x2f, 0x61, 0x6e, 0x79, 0x74, 0x68, 0x69, 0x6e, 0x67, 0x22, 0x2c, 0x22, 0x76, 0x69, 0x73, 0x69, 0x62, 0x69, 0x6c, 0x69, 0x74, 0x79, 0x22, 0x3a, 0x22, 0x57, 0x4f, 0x52, 0x4c, 0x44, 0x22, 0x7d},
        },
    },
    Icto:           "",
    Id:             "",
    IpRestrictions: api.IpRestrictions{},
    Status:         api.Status{},
    Subscriptions:  {
        {
            union: {0x7b, 0x22, 0x62, 0x61, 0x73, 0x65, 0x50, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x65, 0x6e, 0x69, 0x2f, 0x64, 0x69, 0x73, 0x74, 0x72, 0x2f, 0x76, 0x31, 0x22, 0x2c, 0x22, 0x74, 0x79, 0x70, 0x65, 0x22, 0x3a, 0x22, 0x61, 0x70, 0x69, 0x22, 0x7d},
        },
    },
    Zone: "dataplane1",
}
---

[Rover Mapper MapExposures must map exposures correctly - 1]
&api.Rover{
    Authentication: api.Authentication{},
    Exposures:      {
        {
            union: {0x7b, 0x22, 0x61, 0x70, 0x70, 0x72, 0x6f, 0x76, 0x61, 0x6c, 0x22, 0x3a, 0x22, 0x53, 0x49, 0x4d, 0x50, 0x4c, 0x45, 0x22, 0x2c, 0x22, 0x62, 0x61, 0x73, 0x65, 0x50, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x65, 0x6e, 0x69, 0x2f, 0x64, 0x69, 0x73, 0x74, 0x72, 0x2f, 0x76, 0x31, 0x22, 0x2c, 0x22, 0x74, 0x79, 0x70, 0x65, 0x22, 0x3a, 0x22, 0x61, 0x70, 0x69, 0x22, 0x2c, 0x22, 0x75, 0x70, 0x73, 0x74, 0x72, 0x65, 0x61, 0x6d, 0x22, 0x3a, 0x22, 0x68, 0x74, 0x74, 0x70, 0x73, 0x3a, 0x2f, 0x2f, 0x68, 0x74, 0x74, 0x70, 0x62, 0x69, 0x6e, 0x2e, 0x6f, 0x72, 0x67, 0x2f, 0x61, 0x6e, 0x79, 0x74, 0x68, 0x69, 0x6e, 0x67, 0x22, 0x2c, 0x22, 0x76, 0x69, 0x73, 0x69, 0x62, 0x69, 0x6c, 0x69, 0x74, 0x79, 0x22, 0x3a, 0x22, 0x57, 0x4f, 0x52, 0x4c, 0x44, 0x22, 0x7d},
        },
    },
    Icto:           "",
    Id:             "",
    IpRestrictions: api.IpRestrictions{},
    Status:         api.Status{},
    Subscriptions:  nil,
    Zone:           "",
}
---

[Rover Mapper MapSubscriptions must map subscriptions correctly - 1]
&api.Rover{
    Authentication: api.Authentication{},
    Exposures:      nil,
    Icto:           "",
    Id:             "",
    IpRestrictions: api.IpRestrictions{},
    Status:         api.Status{},
    Subscriptions:  {
        {
            union: {0x7b, 0x22, 0x62, 0x61, 0x73, 0x65, 0x50, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x65, 0x6e, 0x69, 0x2f, 0x64, 0x69, 0x73, 0x74, 0x72, 0x2f, 0x76, 0x31, 0x22, 0x2c, 0x22, 0x74, 0x79, 0x70, 0x65, 0x22, 0x3a, 0x22, 0x61, 0x70, 0x69, 0x22, 0x7d},
        },
    },
    Zone: "",
}
---

[Rover Mapper MapRoverResponse must map a Rover to a RoverResponse correctly - 1]
{
 "exposures": [
  {
   "approval": "SIMPLE",
   "basePath": "/eni/distr/v1",
   "type": "api",
   "upstream": "https://httpbin.org/anything",
   "visibility": "WORLD"
  }
 ],
 "id": "eni--hyperion--rover-local-sub",
 "name": "rover-local-sub",
 "status": {
  "processingState": "done",
  "state": "complete",
  "time": "<Any value>"
 },
 "subscriptions": [
  {
   "basePath": "/eni/distr/v1",
   "type": "api"
  }
 ],
 "zone": "dataplane1"
}
---

[Rover Mapper MapRoverResponse must return an error if the input rover is nil - 1]
api.RoverResponse{}
---
